### PR TITLE
First implementation of the subtrees

### PR DESCRIPTION
**Pull request by GingerYouth**

### CODE REVIEW

File: src/main/java/org/cqfn/astranaut/core/algorithms/Subtree.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+    private final Map<Node, List<Integer>> mapping;
+
+    /**
+     * Constructor.
+     *
+     * @param root The root of the subtree
+     * @param nodes The set of nodes in this subtree
+     */
+    @SuppressWarnings("PMD.ConstructorOnlyInitializesOrCallOtherConstructors")
+    public Subtree(final Node root, final Set<Node> nodes) {
```
Comment (kniazkov) [inline src/main/java/org/cqfn/astranaut/core/algorithms/Subtree.java:61]: I think here should be only one parameter, namely, source tree.
Don't perform calculations in constructors.
Expected interface should be like this:

```
Subtree subtree = new Subtree(source);
Node first = subtree.create(Arrays.asList(...)); // first subtree created from the source node and set of nodes
Node second = subtree.create(Arrays.asList(...));
```
Thus, using one subtree object, we should be able to create one or more subtrees

File: src/main/java/org/cqfn/astranaut/core/algorithms/Subtree.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    public SubtreeNode getRoot() {
+        return this.root;
+    }
+
+    /**
+     * Get subtree mapping.
+     *
+     * @return The mapping for this subtree
+     */
+    public Map<Node, List<Integer>> getMapping() {
```
Comment (kniazkov) [inline src/main/java/org/cqfn/astranaut/core/algorithms/Subtree.java:82]: Why not to use package-private (without any modifier) access method?

File: src/main/java/org/cqfn/astranaut/core/algorithms/SubtreeNode.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import org.cqfn.astranaut.core.Fragment;
+import org.cqfn.astranaut.core.Node;
+import org.cqfn.astranaut.core.Type;
+
+/**
+ * Subtree node.
+ *
+ * @since 1.1.4
+ */
+public class SubtreeNode implements Node {
```
Comment (kniazkov) [inline src/main/java/org/cqfn/astranaut/core/algorithms/SubtreeNode.java:36]: I think it should not be public

**Review by kniazkov — CHANGES_REQUESTED** [review]

File: src/main/java/org/cqfn/astranaut/core/algorithms/Subtree.java (hunk 1 above)
Comment (GingerYouth) [reply src/main/java/org/cqfn/astranaut/core/algorithms/Subtree.java]: Changed

File: src/main/java/org/cqfn/astranaut/core/algorithms/Subtree.java (hunk 2 above)
Comment (GingerYouth) [reply src/main/java/org/cqfn/astranaut/core/algorithms/Subtree.java]: Changed

File: src/main/java/org/cqfn/astranaut/core/algorithms/SubtreeNode.java (hunk 3 above)
Comment (GingerYouth) [reply src/main/java/org/cqfn/astranaut/core/algorithms/SubtreeNode.java]: Changed

**Review by kniazkov — APPROVED** [review]

**Review by Darfiro — APPROVED** [review]

**Comment by kniazkov** [comment]
@rultor release this shit, tag is `1.1.4`

**Pull request merged**